spring-integration-sftp/src/main/java/org/springframework/integration/sftp/session/DefaultSftpSessionFactory.java

```diff
@@ -349,57 +355,39 @@ public void setAllowUnknownKeys(boolean allowUnknownKeys) {
 
 	@Override
 	public SftpSession getSession() {
-		Assert.hasText(this.host, "host must not be empty");
-		Assert.hasText(this.user, "user must not be empty");
-		Assert.isTrue(StringUtils.hasText(this.userInfoWrapper.getPassword()) || this.privateKey != null,
-				"either a password or a private key is required");
+		JSchSessionWrapper jschSession = this.sharedJschSession;
+		SftpSession sftpSession;
+		if (this.sharedSessionLock != null) {
+			this.sharedSessionLock.lock();
+		}
 		try {
-			JSchSessionWrapper jschSession;
-			SftpSession sftpSession;
-			if (this.isSharedSession) {
-				this.sharedSessionLock.readLock().lock();
-				try {
-					if (this.sharedJschSession == null || !this.sharedJschSession.isConnected()) {
-						this.sharedSessionLock.readLock().unlock();
-						this.sharedSessionLock.writeLock().lock();
-						try {
-							if (this.sharedJschSession == null || !this.sharedJschSession.isConnected()) {
-								this.sharedJschSession = new JSchSessionWrapper(initJschSession());
-								try {
-									this.sharedJschSession.getSession().connect();
-								}
-								catch (JSchException e) {
-									throw new IllegalStateException("failed to connect", e);
-								}
-							}
-						}
-						finally {
-							this.sharedSessionLock.readLock().lock();
-							this.sharedSessionLock.writeLock().unlock();
-						}
-					}
-					jschSession = this.sharedJschSession;
-					sftpSession = new SftpSession(jschSession);
-					sftpSession.connect();
-				}
-				finally {
-					this.sharedSessionLock.readLock().unlock();
-				}
-			}
-			else {
+			if (jschSession == null || !jschSession.isConnected()) {
 				jschSession = new JSchSessionWrapper(initJschSession());
-				sftpSession = new SftpSession(jschSession);
-				sftpSession.connect();
 			}
-			jschSession.addChannel();
-			return sftpSession;
+			sftpSession = new SftpSession(jschSession);
+			sftpSession.connect();
 		}
 		catch (Exception e) {
 			throw new IllegalStateException("failed to create SFTP Session", e);
 		}
+		finally {
+			if (this.sharedSessionLock != null) {
+				this.sharedSessionLock.unlock();
+			}
+		}
+		if (this.isSharedSession) {
+			this.sharedJschSession = jschSession;
```

Comment:
Why write it back to the field on every call?

In any case, it must be done while the lock is held; otherwise we're back to the possibility of a concurrent update.